controllers/openstackcluster_controller.go

```diff
@@ -627,3 +617,77 @@ func handleUpdateOSCError(openstackCluster *infrav1.OpenStackCluster, message er
 	openstackCluster.Status.FailureReason = &err
 	openstackCluster.Status.FailureMessage = pointer.String(message.Error())
 }
+
+// validateSubnets validates if the amount of IPv4 and IPv6 subnets is allowed.
+func validateSubnets(subnets []infrav1.Subnet) error {
+	isIPv4 := false
+	isIPv6 := false
+
+	for _, subnet := range subnets {
+		ip, _, err := net.ParseCIDR(subnet.CIDR)
+		if err != nil {
+			return err
+		}
+		netIP := net.ParseIP(ip.String())
+		if netIP != nil {
+			if netIP.To4() != nil {
+				if !isIPv4 {
+					isIPv4 = true
+				} else {
+					return fmt.Errorf("multiple IPv4 Subnet not allowed on OpenStackCluster")
+				}
+			} else {
+				if !isIPv6 {
+					isIPv6 = true
+				} else {
+					return fmt.Errorf("multiple IPv6 Subnet not allowed on OpenStackCluster")
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// filterSubnets retrieves the subnets based on the Subnet filters specified on OpenstackCluster.
+func filterSubnets(networkingService *networking.Service, openStackCluster *infrav1.OpenStackCluster) ([]infrav1.Subnet, error) {
+	var subnets []infrav1.Subnet
+	openStackClusterSubnets := openStackCluster.Spec.Subnets
+	if len(openStackClusterSubnets) == 0 {
+		filteredSubnet, err := networkingService.GetNetworkSubnetByFilter(openStackCluster.Status.Network.ID, &infrav1.SubnetFilter{})
```

Comment:
I know `GetNetworkSubnetByFilter` returns a unique subnet, but how this would work with Dual Stack then?
I understand that a dual stack user would specify the subnets via `OpenStackCluster.Spec.Subnets` but what if they don't and rely on `OpenStackCluster.Spec.Network` instead (and let's say the network is dual stack and has 2 subnets).

If we don't want to go down that path, we need to put comments in the code to explain that and why.

Reply:
Good point! I updated to allow 2 subnets when subnets is not specified in the OpenStackCluster spec